rclpy/rclpy/qos.py

```diff
@@ -39,39 +44,43 @@ class QoSProfile:
     def __init__(self, **kwargs):
         assert all('_' + key in self.__slots__ for key in kwargs.keys()), \
             'Invalid arguments passed to constructor: %r' % kwargs.keys()
+
+        if not len(_qos_default_value_for_init):
+            # Any of the setters, upon receiving these None values, would assert.
+            # This can't happen though, because this is only used at definition time, when
+            # we are initializing fully-defined preset profiles.
+            from_profile = Namespace(**{
```

Comment:
It feels dirty that we're relying on an instance of `QoSProfile` in the constructor for `QoSProfile` and that we're essentially mocking it to fake that it exists so we can construct the initial instance. The only options around this I can think of though would be to switch to a Factory pattern, which would be ton of work, or creating a new function in `_rclpy` that would return a more raw representation of the policy, like in the form of a dict, instead of relying on the `QoSProfile` class.

Reply:
Like you said, the only way we expose this information to the python layer is via this QoSProfile class

Reply:
I think it's worth changing the implementation in `_rclpy` so it's not returning a `QoSProfile`. It has been brought up that the current implementation is less than desirable. I've created an issue https://github.com/ros2/rclpy/issues/358 that we can hopefully pick up after in the near future.